pip/pip-418.md

```diff
@@ -0,0 +1,174 @@
+# PIP-418: Determine the behaviors for components that rely on BookKeeper when BookKeeper is not used
+
+# Background knowledge
+
+`PulsarService` has a method for other classes to access the BookKeeper client (`BookKeeper`):
+
+```java
+    public BookKeeper getBookKeeperClient() {
+        ManagedLedgerStorageClass defaultStorageClass = getManagedLedgerStorage().getDefaultStorageClass();
+        if (defaultStorageClass instanceof BookkeeperManagedLedgerStorageClass bkStorageClass) {
+            return bkStorageClass.getBookKeeperClient();
+        } else {
+            // TODO: Refactor code to support other than default bookkeeper based storage class
+            throw new UnsupportedOperationException("BookKeeper client is not available");
+        }
+    }
+```
+
+The BookKeeper client is obtained from the `BookkeeperManagedLedgerStorageClass` interface, which extends the `ManagedLedgerStorageClass` interface. This interface represents the actual storage layer of `ManagedLedgerStorage`.
+
+```java
+    Collection<ManagedLedgerStorageClass> getStorageClasses();
+```
+
+The following components share the same `BookKeeper` instance:
+- `PulsarCompactionServiceFactory`: The default compaction service configured by `compactionServiceFactoryClassName`
+- `EventTimeCompactionServiceFactory`: Another built-in compaction service introduced by PIP-215
+- `StrategicTwoPhaseCompactor`: The compactor used only for the `loadbalancer-service-unit-state` topic (introduced by PIP-352)
+
+This `BookKeeper` instance is also used in:
+- `AdminResource#bookKeeper`: Handles the `/bookies/all` REST API
+- `PersistentTopic#getInternalStats`: Uses the `BookKeeper` object to get a list of ledgers
+
+`PulsarService` also has another method to create a new BookKeeper client from the factory class (`BookKeeperClientFactory`):
+
+```java
+    public BookKeeperClientFactory getBookKeeperClientFactory() {
+        return bkClientFactory;
+    }
+```
+
+The following components leverages this factory to create their own `BookKeeper` instances:
+- `BookkeeperSchemaStorageFactory`: The default schema registry storage configured by `schemaRegistryStorageClassName`
+- `BookkeeperBucketSnapshotStorage`: A non-default built-in delayed delivery tracker configured by `delayedDeliveryTrackerFactoryClassName`
```

Comment:
I’m not entirely sure why separate BookKeeper instances are used for these two cases. While this doesn’t directly affect the goal of the proposal, should we consider addressing or revisiting this as part of a broader cleanup?